Expand documentation for how to build the project

mttkay opened the pull request:
I expanded on the documentation for:

- Which tools are required to work with this repo.
- How to use them to build the repo.

mttkay commented:
@eliias WDYT? The tooling setup is a bit messy. I wonder if we should rely purely on a single top-level `.tool-versions`, or perhaps even better, provide a build container? That could be a future iteration, for now I'm just looking for things to switch from red to green!

mttkay commented:
We should also consider dropping `.ruby-version`; asdf subsumes this and I'm not sure how many different environment managers we need to support and maintain?

mttkay commented on gems/yrb-actioncable/.tool-versions:
```diff
@@ -1 +1,4 @@
 redis 6.2.7
+ruby 3.2.0
+nodejs 19.6.0
```
I wasn't totally sure if we needed to pin node as well to get `npm` or if adding `yarn` is sufficient :thinking:

eliias replied:
This is fine. We should pin the Node.js version in `.nvmrc` and the Ruby one in `.ruby-version` to ensure things keep working for non `asdf` users.

mttkay replied:
OK :+1: 

btw I noticed that Turborepo says you should choose either npm, yarn or pnpm. Do we actually need both? :thinking: 

https://turbo.build/repo/docs/getting-started/create-new#which-package-manager-do-you-want-to-use

mttkay commented:
I also get a dirty `package.json` when running the example app, which triggers `build:js`:

```diff
diff --git a/examples/collaborative-text-editor/package.json b/examples/collaborative-text-editor/package.json
index ed2350a..b8571df 100644
--- a/examples/collaborative-text-editor/package.json
+++ b/examples/collaborative-text-editor/package.json
@@ -10,7 +10,7 @@
     "@tiptap/extension-collaboration": "^2.0.0-beta.209",
     "@tiptap/extension-collaboration-cursor": "^2.0.0-beta.209",
     "@tiptap/starter-kit": "^2.0.0-beta.209",
-    "@y-rb/actioncable": "*",
+    "@y-rb/actioncable": "^0.1.5",
     "esbuild": "^0.16.10",
     "lib0": "^0.2.58",
     "prosemirror-commands": "^1.5.0",
```

is that because I use a different version of yarn?

mttkay commented:
I'm confused about what caused this change; I am not familiar with how node manages packages, but this suggests that `package.json` is manually maintained and not generated from something else? https://docs.npmjs.com/cli/v9/configuring-npm/package-json

eliias commented:
There are a few things here. The `yrb-actioncable` repo is a monorepo, so this might help to understand why there isn't a single root config for everything. e.g., `packages/*` contains the npm modules, `gems` contains the Ruby gems, `examples` is for demo applications. For npm modules, this is a bit easier to solve, because you can have local (private) npm module to share configuration (e.g., see custom eslint rules).

The monorepo is organized via [https://turbo.build/repo](https://turbo.build/repo) and utilizes Yarn workspaces. This automatically de-duplicates node modules across all workspace. You should be able to simply run `yarn` in the root folder to set everything up (let me know if the lock file issue is still prevalent after that, but I think it should be gone).

👍 on adding Ruby/Node/Yarn to the respective `.tool-versions` files. `npm` comes with Node, but we use `yarn`, so adding both should be better.

> We should also consider dropping .ruby-version; asdf subsumes this and I'm not sure how many different environment managers we need to support and maintain?

`.ruby-version` is by far the most common dotfile used to control Ruby versions. I think it is not safe to assume that folks use `asdf`. As `asdf` falls back to `.ruby-version` to determine the correct Ruby version, this seems to be the best compromise to support Ruby only development for rvm, rbenv, chruby users.

> I also get a dirty package.json when running the example app, which triggers build:js:

This is due to how turbo works, we use the `"*"` pattern, as described in the [turborepo guide](https://turbo.build/repo/docs/handbook/sharing-code), to get the latest local version of the npm package. This is particularly useful when a developer updates any code in `@yrb/actioncable`. The changes will be immediately reflected in the `collaborative-text-editor` demo without running `yarn` or `yarn install`. In other words, when you run `yarn turbo build:watch`, the produced JavaScript artifacts will automatically lead to Rails reloading the JavaScript bundle.

```
diff --git a/examples/collaborative-text-editor/package.json b/examples/collaborative-text-editor/package.json
index ed2350a..b8571df 100644
--- a/examples/collaborative-text-editor/package.json
+++ b/examples/collaborative-text-editor/package.json
@@ -10,7 +10,7 @@
     "@tiptap/extension-collaboration": "^2.0.0-beta.209",
     "@tiptap/extension-collaboration-cursor": "^2.0.0-beta.209",
     "@tiptap/starter-kit": "^2.0.0-beta.209",
-    "@y-rb/actioncable": "*",
+    "@y-rb/actioncable": "^0.1.5",
     "esbuild": "^0.16.10",
     "lib0": "^0.2.58",
     "prosemirror-commands": "^1.5.0",
```

eliias commented:
Ah, two more things:
- For commit messages, all the `y-rb` repos follow the pattern described here https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit. This allows for automatic release management and changelog generatino (determine major, minor, patch version).
- I'd pin to the latest Node.js LTS instead of the latest, which is `18.14.0`

mttkay commented:
Thanks for the detailed explanations!

> The monorepo is organized via https://turbo.build/repo and utilizes Yarn workspaces.

Ah thank you, I did not know -- sorry for stumbling around this, the last time I even touched JS and related tooling was probably 10 years ago and I'm pretty sure yarn wasn't even a thing then, much less Turborepo :grimacing: I have some catching up to do.

I will add this explanation to the README if you don't mind; I don't think we can assume that a newcomer to this repository will know what Turborepo is and how it affects the build process.

> There are a few things here. The yrb-actioncable repo is a monorepo, so this might help to understand why there isn't a single root config for everything.
> ...
> You should be able to simply run yarn in the root folder to set everything up

I'm not sure how to reconcile these two ideas: either we want to be able to build everything from the project root, which means we need a top-level `.tool-versions`/`.ruby-version`/`.nvmrc` etc. Otherwise running `yarn` will just fail unless it's installed globally, which I don't think we should assume. Or, we provide a top-level build script that dives into subfolders and runs `yarn`, bundler etc. there.

Also, it looks like Turborepo suggests we should use the `turbo build` command to build everything laid out this way: https://turbo.build/repo/docs/handbook/building-your-app

I think this requires installing it globally first:

```
npm install turbo --global
```

It will require adding a top-level `.tool-versions` as well since `turbo` requires `npm`... Lots of turtles stacked on top of each other :grimacing: 

I will see if I can produce something working today and update the PR accordingly.

> For commit messages, all the y-rb repos follow the pattern described here https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit. This allows for automatic release management and changelog generation (determine major, minor, patch version). The package used for that is https://github.com/changesets/changesets

My bad -- I'll make sure to review the contribution guidelines :+1: 

> I'd pin to the latest Node.js LTS instead of the latest, which is 18.14.0

Will do!

mttkay commented on .tool-versions:
```diff
@@ -0,0 +1,4 @@
+redis 6.2.7
+ruby 3.2.0
+nodejs 18.14.0
+yarn 1.22.19
```
We need this top-level definition since these tools are required to build the repo modules with `turbo` (though not sure if Ruby is actually needed at the top-level)

mttkay commented on README.md:
```diff
@@ -8,18 +8,51 @@
 
-This is a monorepo with [npm modules](./packages) and [Ruby gems](./gems).
-Please check the respective sub-repos for more information. 
+TODO: Explain what this library does.
```
@eliias I added some TODOs. This is where I need your help but I wanted to provide extension points already. We should tackle these in follow-up PRs. The most important thing to understand when coming to a repository is why it exists in the first place.

mttkay commented on examples/collaborative-text-editor/Procfile.dev:
```diff
@@ -1,4 +1,4 @@
 js: yarn build:js --watch
 css: bin/rails tailwindcss:watch
-redis: redis-server
+redis: docker run --rm --name redis -p 6379:6379 redis:6-alpine
```
TODO: this needs to mount a tmpdir so redis can dump data. Still need to test this.

eliias commented:
> I will add this explanation to the README if you don't mind; I don't think we can assume that a newcomer to this repository will know what Turborepo is and how it affects the build process.

Yes please. Thanks 🙌 

> I'm not sure how to reconcile these two ideas: either we want to be able to build everything from the project root, which means we need a top-level .tool-versions/.ruby-version/.nvmrc etc. Otherwise running yarn will just fail unless it's installed globally, which I don't think we should assume. Or, we provide a top-level build script that dives into subfolders and runs yarn, bundler etc. there.

Yeah, this is tricky. I am not sure what is best here. From the `turbo` perspective, having `Node.js/yarn` on the path should be enough, as `turbo` is a root-level dependency (see `package.json` in root). That said, the Ruby `gem` isn't included in the `turbo` setup, because it is mostly a JavaScript ecosystem tool.

I think we could make it work, but then we would need to make sure that Ruby is available + `bundle install` runs in all Ruby sub-repos, which isn't a feature `turbo` offers for Ruby. Thinking ahead, the same is true for automated releases, which is only done for `npm` packages (see GitHub releases, which are created automatically). The current state kinda assumes that Ruby devs know what to do in a `gem` and the JavaScript part is heavily automated.

> It will require adding a top-level .tool-versions as well since turbo requires npm... Lots of turtles stacked on top of each other 😬

I think this should just work right now, as there is already a `.ruby-version` and `.nvmrc` file at the root level which `asdf` should pick up properly?

> My bad -- I'll make sure to review the contribution guidelines 👍

I'll add a section to the README, as I also want to make sure it is easy to understand how to create a changeset and an automated release.

eliias commented:
Hm… the more I think about it, the more I agree with your original statement about root level `.nvmrc`, `.ruby-version` and `.tool-versions` files. I also believe we might want to maintain both, `.tool-versions` and `.ruby-version` to avoid friction (see https://github.com/asdf-vm/asdf-ruby#ruby-version-file).

eliias commented:
@mttkay check https://github.com/y-crdt/yrb-actioncable/pull/27 for reference.

mttkay commented:
Yeah I was going to say, I don't think `asdf` falls back to these tool-specific files by default. 

With `ruby` pinned in `.tool-versions`:

```shell
[13:17:33] work/yrb-actioncable::mk-add-ruby-to-tool-versions ✗ asdf which ruby
/home/mk/.asdf/installs/ruby/3.2.0/bin/ruby
```

Dropping `ruby` entry:

```shell
[13:17:14] work/yrb-actioncable::mk-add-ruby-to-tool-versions ✗ asdf which ruby
/home/mk/.asdf/installs/ruby/3.0.5/bin/ruby
```

Which is my global default Ruby currently, not what's specified in `.ruby-version`.

Thanks for the proposal -- I'll test your PR :+1:

mttkay commented:
I rebased my changed onto your @eliias --thanks for helping out to disentangle this!

The branch name does not make sense anymore because all that's left in this PR are the changes to the README file, but I renamed the PR to reflect this.

Could you have another look?

eliias commented on README.md:
````diff
@@ -8,19 +8,50 @@
+
+This is a monorepo built and laid out with [turborepo](https://turbo.build).
+The following dependencies need to exist in order to build this repository
+with `turbo` and run the sample application:
+
+- `Ruby`
+- `yarn`
+- `Node.js/npm`
+
+Refer to the respective configuration files to see which versions are required.
+The easiest option is to use `asdf` to install and manage these tools, but you
+may also refer to `/.tool-versions` and install these manually on your host machine.
+
+## Build
 
-Install gem and npm package:
+Run the following commands from the top-level directory of this repository.
 
+Install `turbo`:
+
+```shell
+$ npm install turbo --global
 ```
-gem install y-rb_actioncable
-yarn add @y-rb/actioncable
+
+Install JavaScript dependencies:
+
+```shell
+$ yarn
 ```
 
-### Example
+Build the repository:
+
+```shell
+$ turbo build
+```
+
+## Usage
````
@mttkay sorry for not coordinating better. I have already added a development specific section now https://github.com/y-crdt/yrb-actioncable#development to help with onboarding specifically for contributors. There are a few differences:
- We don't need to install turbo manually, it will just be there when running `yarn` the first time. `yarn` is capable of resolving binaries from the `.bin` folder in the `node-modules` directory, so a global installation isn't necessary (and we can control the turbo version).
- I kept the usage instructions, because they are important for actual users
- Added build & deploy information (including bumping the version, changesets and information about GitHub Actions used).
- Haven't added a top-level description for all packages (for me that was covered by _Usage_), but I will take what you have in your PR and add it.

mttkay replied:
OK, thanks this is great. Just a not that I did not remove the Usage section; the diff just pushed it below the `Setup` section.

Appreciate the help!

mttkay replied:
Should I close this then?

eliias replied:
@mttkay I would just take this PR, rebase from main, fix the few conflicts and merge than. Ok?

mttkay replied:
Yes sure, thanks :+1:

eliias replied:
> the diff just pushed it below the Setup section.

🤦 missed that.
